smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+                            || propertyNames.contains(memberShape.getMemberName())) {
+                        memberShapeToPropertyName.put(memberShape.getId(), getPropertyTraitName(memberShape)
+                            .orElse(memberShape.getMemberName()));
+                    }
+                }
+                // nesting is taking place, so index top level input/output members as not property.
+                if (!inputPropertiesShape.getId().equals(operationShape.getInputShape())) {
+                    for (MemberShape memberShape : model.expectShape(operationShape.getInputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+
+                Shape outputPropertiesShape = getOutputPropertiesShape(operationShape);
+                operationToOutputPropertiesShape.put(operationShapeId, outputPropertiesShape.getId());
+                for (MemberShape memberShape : outputPropertiesShape.members()) {
+                    if (identifierIndex.getOperationOutputBindings(resourceShape, operationShape).values()
+
+                            .contains(memberShape.getMemberName())) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    } else {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(),
+                                doesNotRequireProperty(memberShape));
+                    }
+                    if (doesMemberShapeRequireProperty(memberShape.getId())
+                            || propertyNames.contains(memberShape.getMemberName())) {
+                        memberShapeToPropertyName.put(memberShape.getId(), getPropertyTraitName(memberShape)
+                            .orElse(memberShape.getMemberName()));
+                    }
+                }
+                // nesting is taking place, so index top level input/output members as not property.
+                if (!outputPropertiesShape.getId().equals(operationShape.getOutputShape())) {
+                    for (MemberShape memberShape : model.expectShape(operationShape.getOutputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+            }
+        }
+    }
+
+    public static PropertyBindingIndex of(Model model) {
+        return model.getKnowledge(PropertyBindingIndex.class, PropertyBindingIndex::new);
+    }
+
+    /**
+     * Gets the property name for a given member shape. Returns empty optional if the
+     * member shape does not correspond to a property.
+     *
+     * @param memberShapeId the ShapeId of the member shape to get the property name for.
+     * @return the property name for a given member shape if there is one.
+     */
+    public Optional<String> getPropertyName(ShapeId memberShapeId) {
```

Comment:
How does this not need to also know the resource? What if a shape is reused by more than one resource?

Reply:
It's impossible for a more than one resource to use a shape and try to implement different property names for the same members.